feat(ssr): support external true (#10939)
vitejs/vite: docs/config/ssr-options.md

## ssr.external

- **Type:** `string[] | true`
- **Related:** [SSR Externals](/guide/ssr#ssr-externals)

> Externalize dependencies for SSR. By default, all dependencies are externalized, except linked dependencies for HMR. This can opted-out by adding the dependency to this option, or set `true` to force externalize all dependencies.

Reviewer: it took me a little while to understand the opt-out was referring to linked dependencies. I might reword this a bit:
```suggestion
Externalize the given packages and their transitive dependencies for SSR. `true` will include all packages. By default, all dependencies are externalized except for linked dependencies. This option takes priority over `ssr.noExternal`.
```

## ssr.noExternal

- **Type:** `string | RegExp | (string | RegExp)[] | true`
- **Related:** [SSR Externals](/guide/ssr#ssr-externals)

> Prevent listed dependencies from being externalized for SSR. If `true`, no dependencies are externalized. This option has a higher priority than `ssr.external` except if a dependency is explicitly specified in the `ssr.external` array.

Reviewer: The new sentence is not clear to me

Let's say a depends on b depends on c.

If a and b are in noExternal and b is in external, which are externalized?
If a is in noExternal and b is in external, which are externalized?

Author: > If a and b are in noExternal and b is in external, which are externalized?

If `a` and `b` are explicitly configured in `external` and `noExternal` configs, `a` will be noExternalized, and `b` will be externalized.

If `a` and `b` are explicitly configured in `noExternal` configs, `external` config is unconfigured because `b` is already external by default, `a` and `b` will be noExternalized. 

> If a is in noExternal and b is in external, which are externalized?

`a` will be noExternalized, `b` will be externalized.

---

The priority order it something like, (highest to lowest priority) explicit external > explicit noExternal > (is linked dep ? default noExternal : default external).

Author: ```suggestion
Prevent listed dependencies from being externalized for SSR. If `true`, no dependencies are externalized. If a dependency is explicitly defined in `ssr.external`, it will take priority and be externalized instead.
```

maybe something like this is clearer?

Reviewer: just to make sure I understand, can you clarify what will happen to `c` in my two examples?

Author: in both examples because `b` is externalized, `c` will also be externalized as Vite can't touch `b`'s code.

> If `a` and `b` are explicitly configured in `noExternal` configs, `external` config is unconfigured because `b` is already external by default, `a` and `b` will be noExternalized.

For this case where `b` is `noExternalized`, and `c` isn't explicitly configured anywhere, `c` will also be noExternalized otherwise it would error in runtime in strict dependency installations, because your app in root can't import `c`.

Reviewer: yeah, I like your new suggested version of the docs. just one minor tweak I'd make to it:
```suggestion
Prevent listed dependencies from being externalized for SSR. If `true`, no dependencies are externalized. If a dependency is explicitly defined in `ssr.external`, it will take priority and be externalized along with its transitive dependencies.
```

> Prevent listed dependencies from being externalized for SSR. If `true`, no dependencies are externalized. If a dependency is explicitly defined in `ssr.external`, it will take priority and be externalized along with its transitive dependencies.

Reviewer: we should clarify that this works both with packages names and with direct entries. also the part about ssr.external applying transitively should probably go in the section for ssr.external and the part about linked packages should probably go here, so I've swapped those in the suggestions
```suggestion
Bundles the given packages or modules within a package for SSR. If `true`, no dependencies are externalized. Linked dependencies will be bundled by default in order to support HMR on those packages.
```
